### ISSUE TITLE
[Bug] Routes: `redirect` does `pushState` instead of `replaceState`, breaking browser back buttons

### ISSUE THREAD

**Issue by tonywu6** [umijs/umi#8703]
## What happens?

`redirect` routes now update window locations by doing a `window.history.pushState` instead of `window.history.replaceState`. **This renders browsers' back buttons useless** as trying to go to the previous page lands us in the location with `redirect`, bouncing us back to the most recent page.

## Minimal Reproducible Example

https://github.com/tonywu6/umi-4-issue-redirect

## How To Reproduce

1. Create a minimal umi app; in app config, in `routes`, provide the following

```js
  routes: [
    { path: '/', component: 'index' },
    {
      path: '/lorem',
      routes: [
        {
          path: '',
          redirect: 'ipsum',
        },
        {
          path: 'ipsum',
          routes: [
            {
              path: '',
              redirect: 'dolor',
            },
            {
              path: 'dolor',
              component: 'main',
            },
          ],
        },
      ],
    },
  ],
```

2. Start the dev server, then navigate to `/`;
3. Navigate to `/lorem`;
4. Observe that the page is redirected to `/lorem/ipsum/dolor`;
5. Try returning to `/` using the **back button**;
6. **Observe that it is impossible to return to `/`; observe that the address briefly changes to `/lorem/ipsum`, then changes back to `/lorem/ipsum/dolor`.**

https://user-images.githubusercontent.com/93302820/180778384-47047ef8-9ac9-46bb-a8a8-55d016d3c513.mov

#### Expected behavior

After landing on a page via redirects, I should still be able to visit whatever previous page was using the back button.

#### Current behavior

It is impossible to return to the previous page using the back button.

## Context

- **Umi Version**: `4.0.8`
- **Node Version**: `v14.18.1`
- **Platform**: `Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36`

Previously, with `react-router@5`, redirecting is done by the `<Redirect>` component. With that gone in `react-router@6`, redirecting is instead done by `<Navigate>`:

https://github.com/umijs/umi/blob/591fa7c40523b7753dc746b34f743d952cfc3c67/packages/renderer-react/src/routes.tsx#L55-L57

where `<NavigateWithParams>` is

https://github.com/umijs/umi/blob/591fa7c40523b7753dc746b34f743d952cfc3c67/packages/renderer-react/src/routes.tsx#L38-L45

**The `replace` prop is needed in `<Navigate>` for it to use `replaceState`.**

**Comment by tonywu6** [comment]
See also #8651

**Comment by sorrycc** [comment]
Fixed via #8662
